File: src/python/gudhi/bottleneck.py
# This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
# See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
# Author(s):       Hannah Schreiber
#
# Copyright (C) 2025 Inria
#
# Modification(s):
#   - YYYY/MM Author: Description of the modification

__license__ = "GPL v3"

from collections.abc import Sequence
from numpy.typing import ArrayLike

from gudhi import _bottleneck_ext as t
from gudhi.hera import _diagram_as_numpy_array


def bottleneck_distance(diagram_1: ArrayLike, diagram_2: ArrayLike, e: float = None) -> float:
    """Compute the Bottleneck distance between two diagrams.
    Points at infinity and on the diagonal are supported.

    :param diagram_1: The first diagram.
    :type diagram_1: numpy array of shape (m,2)
    :param diagram_2: The second diagram.
    :type diagram_2: numpy array of shape (n,2)
    :param e: If `e` is 0, this uses an expensive algorithm to compute the
        exact distance.
        If `e` is not 0, it asks for an additive `e`-approximation, and
        currently also allows a small multiplicative error (the last 2 or 3
        bits of the mantissa may be wrong). This version of the algorithm takes
        advantage of the limited precision of `double` and is usually a lot
        faster to compute, whatever the value of `e`.
        Thus, by default (`e=None`), `e` is the smallest positive double.
    :type e: float
    :rtype: float
    :returns: the bottleneck distance.
    """
    return t._bottleneck_distance(
        _diagram_as_numpy_array(diagram_1), _diagram_as_numpy_array(diagram_2), e
    )

File: src/python/gudhi/_bottleneck.cc
/*    This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
 *    See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
 *    Author(s):       Marc Glisse
 *
 *    Copyright (C) 2020 Inria
 *
 *    Modification(s):
 *      - 2025/01 Vincent Rouvreau: Use nanobind instead of PyBind11 for python bindings
 *      - 2025/04 Hannah Schreiber: Re-add possibility of native Python sequences as input
 *      - YYYY/MM Author: Description of the modification
 */

#include <limits>
#include <optional>
#include <utility>  //std::pair

#include <nanobind/nanobind.h>
#include <nanobind/stl/optional.h>

#include <gudhi/Bottleneck.h>
#include <python_interfaces/diagram_utils.h>

namespace nb = nanobind;

// Indices are added internally in bottleneck_distance, they are not needed in the input.
static auto _make_point(double x, double y, std::size_t) { return std::pair(x, y); };

// For compatibility with older versions, we want to support e=None.
double _bottleneck(const Tensor_dgm& d1, const Tensor_dgm& d2, std::optional<double> epsilon)
{
  double e = epsilon.value_or((std::numeric_limits<double>::min)());
  auto d1_view = d1.view(); // views have to live until the end of the method, as not copied in
  auto d2_view = d2.view(); // the iterator of array_to_range_of_pairs
  // I *think* the call to request() in array_to_range_of_pairs has to be before releasing the GIL.
  auto diag1 = array_to_range_of_pairs(d1_view, _make_point);
  auto diag2 = array_to_range_of_pairs(d2_view, _make_point);

  nb::gil_scoped_release release;

  return Gudhi::persistence_diagram::bottleneck_distance(diag1, diag2, e);
}

NB_MODULE(_bottleneck_ext, m)
{
  m.attr("__license__") = "GPL v3";
  m.def("_bottleneck_distance",
        &_bottleneck,
        nb::arg("diagram_1"),
        nb::arg("diagram_2"),
        nb::arg("e") = nb::none());
}
